fingerprint2.js

```diff
@@ -209,6 +216,94 @@
       }
       return keys
     },
+    isFlashKey: function (keys){
+      if (!this.options.excludeIsFlash){
+          keys.addPreprocessedComponent({key: 'is_flash', value:this.getIsFlash()})
+      }
+      return keys
+    },
+    getIsFlash: function(){
+      var objPlugin = navigator.plugins["Shockwave Flash"];
+      if (objPlugin) {
+        return true;
+      }
+      return false;
+    },
+    flashVersionKey: function(keys){
+       if (!this.options.excludeFlashVersion){
+          keys.addPreprocessedComponent({key: 'flash_version', value:this.getFlashVersion()})
+      }
+      return keys
+    },
+    getFlashVersion: function() {
+      if (this.getIsFlash()) {
+        var objPlayerVersion = swfobject.getFlashPlayerVersion();
+        return objPlayerVersion.major + "." + objPlayerVersion.minor + "." + objPlayerVersion.release;
+      }
+      return "unknown";
+    },
+    isSilverlightKey: function(keys){
+      if (!this.options.excludeIsSilverlight){
+          keys.addPreprocessedComponent({key: 'is_silverlight', value:this.getIsSilverlight()})
+      }
+      return keys
+    },
+    getIsSilverlight: function() {
+      var objPlugin = navigator.plugins["Silverlight Plug-In"];
+      if (objPlugin) {
+        return true;
+      }
+      return false;
+    },
+    silverlightVersionKey: function(keys){
+      if (!this.options.excludeSilverlightVersion){
+          keys.addPreprocessedComponent({key: 'silverlight_version', value:this.getSilverlightVersion()})
+      }
+      return keys
+    },
+    getSilverlightVersion: function() {
+      if (this.getIsSilverlight()) {
+        var objPlugin = navigator.plugins["Silverlight Plug-In"];
+        return objPlugin.description;
+      }
+      return "unknown";
+    },
+    isMimeTypesKey: function(keys){
+      if (!this.options.excludeIsMimeTypes){
+          keys.addPreprocessedComponent({key: 'is_mime_types', value:this.getIsMimeTypes()})
+      }
+      return keys
+    },
+    getIsMimeTypes: function() {
+      if (navigator.mimeTypes.length) {
+        return true;
+      }
+      return false;
+    },
+    mimeTypesKey: function (keys){
+      if (!this.options.excludeMimeTypes){
+          keys.addPreprocessedComponent({key: 'mime_types', value:this.getMimeTypes()})
+      }
+      return keys
+    },
+    getMimeTypes: function() {
+      var mimeTypeList = "";
+
+      for (var i = 0; i < navigator.mimeTypes.length; i++) {
+        if (i == navigator.mimeTypes.length - 1) {
+          mimeTypeList += navigator.mimeTypes[i].description;
+        } else {
+          mimeTypeList += navigator.mimeTypes[i].description + ", ";
+        }
+      }
+      return mimeTypeList;
+    },
+    isCookieKey: function(keys) {
+      if (!this.options.excludeMimeTypes){
+          keys.addPreprocessedComponent({key: 'is_cookie', value:navigator.cookieEnabled})
```

Comment:
I'm not sure how this can be used to create a fingerprint. It probably changes all the time (depending on adblock, page you're visiting, incognito mode, etc).

Reply:
I think it's not just about generating device fingerprints. It can be used as a separate project for the back-end to determine the degree of risk to the user environment